File: backend/app/models/health_check.py
from __future__ import annotations

from datetime import datetime
from enum import Enum

from sqlalchemy import CheckConstraint, DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class HealthCheckStatus(str, Enum):
    healthy = "healthy"
    unhealthy = "unhealthy"
    timeout = "timeout"
    error = "error"


class HealthCheck(Base):
    __tablename__ = "health_checks"
    __table_args__ = (
        CheckConstraint(
            "status in ('healthy', 'unhealthy', 'timeout', 'error')",
            name="ck_health_checks_status",
        ),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    project_id: Mapped[int] = mapped_column(
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    target_url: Mapped[str] = mapped_column(String(2048), nullable=False)
    status: Mapped[str] = mapped_column(String(32), nullable=False)
    http_status_code: Mapped[int | None] = mapped_column(Integer, nullable=True)
    response_time_ms: Mapped[int | None] = mapped_column(Integer, nullable=True)
    checked_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    response_preview: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

File: frontend/src/types/healthCheck.ts
export type HealthCheckStatus = "healthy" | "unhealthy" | "timeout" | "error";

export interface HealthCheck {
  id: number;
  project_id: number;
  target_url: string;
  status: HealthCheckStatus;
  execution_source?: "manual" | "scheduled";
  http_status_code: number | null;
  response_time_ms: number | null;
  checked_at: string;
  error_message: string | null;
  response_preview: string | null;
  created_at: string;
}

export type HealthMonitorCadence = 15 | 60 | 360 | 1440;

export interface HealthMonitorSchedule {
  project_id: number;
  enabled: boolean;
  cadence_minutes: HealthMonitorCadence;
  next_run_at: string | null;
  last_started_at: string | null;
  last_completed_at: string | null;
  last_outcome: HealthCheckStatus | null;
  consecutive_failures: number;
  created_at: string | null;
  updated_at: string | null;
}

export interface HealthCheckRunInput {
  url?: string;
}
